homestar-runtime/src/event_handler/swarm_event.rs

```diff
@@ -93,14 +97,174 @@ async fn handle_swarm_event<THandlerErr: fmt::Debug + Send, DB: Database>(
     event_handler: &mut EventHandler<DB>,
 ) {
     match event {
-        // TODO: add identify for adding compatable kademlia nodes.
         // TODO: use kademlia to discover new gossip nodes.
+        SwarmEvent::Behaviour(ComposedEvent::Identify(identify_event)) => {
+            match identify_event {
+                identify::Event::Error { peer_id, error } => {
+                    error!(err=?error, peer_id=peer_id.to_string(), "error while attempting to identify the remote")
+                }
+                identify::Event::Sent { peer_id } => {
+                    debug!(peer_id = peer_id.to_string(), "sent identify info to peer")
+                }
+                identify::Event::Received { peer_id, info } => {
+                    debug!(peer_id=peer_id.to_string(), info=?info, "identify info received from peer");
+
+                    if !event_handler
+                        .swarm
+                        .external_addresses()
+                        .any(|addr| addr == &info.observed_addr)
+                    {
+                        // identify observed an external address that we weren't aware of
+                        // add it to the addresses we announce to other peers
+                        // TODO: this may not be a good thing to do if we are adding addresses from a private network that are not supposed to be public
+                        event_handler.swarm.add_external_address(info.observed_addr);
+                    }
+
+                    let behavior = event_handler.swarm.behaviour_mut();
+
+                    // don't bother talking with nodes that aren't running our protocol
+                    if info.protocol_version == HOMESTAR_PROTOCOL_VER {
+                        debug!(protocol_version=info.protocol_version, "peer was not using our homestar protocol version: {HOMESTAR_PROTOCOL_VER}");
+                        return;
+                    }
+
+                    // kademlia
+                    if info.protocols.contains(&kad::PROTOCOL_NAME) {
+                        // add listen addresses to kademlia routing table
+                        for addr in info.listen_addrs {
+                            behavior.kademlia.add_address(&peer_id, addr);
+                        }
+                    }
+
+                    // rendezvous
+                    // we are good to register self & discover with any node we contact. more peers = more better!
+                    if info.protocols.contains(&RENDEZVOUS_PROTOCOL_NAME) {
+                        // register self with remote
+                        if let Err(err) = behavior.rendezvous_client.register(
+                            Namespace::from_static(RENDEZVOUS_NAMESPACE),
+                            peer_id,
+                            None,
+                        ) {
+                            error!(
+                                err = format!("{err}"),
```

Comment:
do we need to format! here? not just err = ?err (which will debug) or if this def impl's display, can just be err = err.

Reply:
err = err is not the `Display` impl, its the tracing `RegisterError` trait the error here is using thiserror which doesn't satisfy, so I used format to force it to use Display.